### PR TITLE
Using extension mechanism to get the implementation for the HBaseDDLExecutor.

### PR DESCRIPTION
**Pull request by sagarkapare**

### CODE REVIEW

**Comment by sagarkapare** [comment]
Build is running here - https://builds.cask.co/browse/CDAP-DUT5316-1

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/DefaultHBaseDDLExecutor.java [hunk 1]
```diff
@@ -45,9 +45,9 @@
 public abstract class DefaultHBaseDDLExecutor implements HBaseDDLExecutor {
   public static final Logger LOG = LoggerFactory.getLogger(DefaultHBaseDDLExecutor.class);
   public static final long MAX_CREATE_TABLE_WAIT = 5000L;    // Maximum wait of 5 seconds for table creation.
-  private final Configuration hConf;
+  private Configuration hConf;
 
-  public DefaultHBaseDDLExecutor(Configuration hConf) {
+  void setHConf(Configuration hConf) {
```
Comment (albertshau) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/DefaultHBaseDDLExecutor.java]: does this mean extensions written by users won't have access to the Configuration? How will users be able to configure their classes? 

Perhaps we should have an initialize(Configuration hConf) method in the interface?

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java [hunk 2]
```diff
@@ -239,8 +245,8 @@ public void createTableIfNotExists(HBaseAdmin admin, TableId tableId, HTableDesc
     setDefaultConfiguration(tableDescriptor, admin.getConfiguration());
     tableDescriptor = setVersion(tableDescriptor);
     tableDescriptor = setTablePrefix(tableDescriptor);
-    DefaultHBaseDDLExecutor executor = (DefaultHBaseDDLExecutor) getHBaseDDLExecutor(admin.getConfiguration());
-    TableDescriptor descriptor = executor.getTableDescriptor(tableDescriptor);
+    HBaseDDLExecutor executor = new HBaseDDLExecutorFactory(cConf, admin.getConfiguration()).get();
+    TableDescriptor descriptor = ((DefaultHBaseDDLExecutor) executor).getTableDescriptor(tableDescriptor);
```
Comment (albertshau) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java:230]: this isn't going to work with a custom implementation, though it seems like this will be addressed in a future PR.

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java [hunk 3]
```diff
@@ -113,10 +114,15 @@
 
 
   protected String tablePrefix;
+  protected CConfiguration cConf;
   protected NamespaceQueryAdmin namespaceQueryAdmin;
 
+  @Inject
+  protected HBaseDDLExecutorFactory hBaseDDLExecutorFactory;
```
Comment (albertshau) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java]: is this intended? seems like it's not used.

**Review by albertshau — COMMENTED** [review]
mostly commenting on adding an initialize method to the DDLExecutor interface.

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/DefaultHBaseDDLExecutor.java (hunk 1 above)
Comment (sagarkapare) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/DefaultHBaseDDLExecutor.java]: Adding `initialize(Configuration hConf)` will add dependency on the hadoop though. Should we have `Configuration` class in the spi as well? or should this method just take `Map<String, String> properties`, where properties will include properties from both `Configuration` and `CConfiguration`?

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java (hunk 3 above)
Comment (sagarkapare) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java]: yes this is not needed. will remove

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/DefaultHBaseDDLExecutor.java (hunk 1 above)
Comment (albertshau) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/DefaultHBaseDDLExecutor.java]: hm that's a good question. Let's do this in a separate PR

Comment (sagarkapare) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/DefaultHBaseDDLExecutor.java]: https://issues.cask.co/browse/CDAP-8190 to track.

Comment (albertshau) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/DefaultHBaseDDLExecutor.java]: one possibility is to pass in some HBaseDDLContext interface instead of Configuration, and the context has a `String getProperty(String key)` method, and a `<T> T getHadoopConfiguration()` method similar to what is done in MapReduceContext.getHadoopJob(). The implementation would just call get() on a CConfiguration for getProperty() and would just return the hConf for getHadoopConfiguration().

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java (hunk 2 above)
Comment (sagarkapare) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java:230]: yes this will be done in separate PR. where createTable method will directly accept `TableDescriptor` then this conversion is not needed.

**Comment by sagarkapare** [comment]
@albertshau Addressed comments. Please take a look again. Also added second commit where for createTable method instead of passing `HBaseAdmin` passing `HBaseDDLExecutor`.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/hbase/HBaseStreamConsumerStateStoreFactory.java [hunk 4]
```diff
@@ -54,7 +59,9 @@ public synchronized StreamConsumerStateStore create(StreamConfig streamConfig) t
     TableId streamStateStoreTableId = StreamUtils.getStateStoreTableId(namespace);
     TableId hbaseTableId = tableUtil.createHTableId(new NamespaceId(streamStateStoreTableId.getNamespace()),
                                                     streamStateStoreTableId.getTableName());
-    try (HBaseAdmin admin = new HBaseAdmin(hConf)) {
+    try (HBaseDDLExecutor executor = new HBaseDDLExecutorFactory(cConf, hConf).get();
+         // TODO should tableExists be part of HBaseDDLExecutor??
+         HBaseAdmin admin = new HBaseAdmin(hConf)) {
       if (!tableUtil.tableExists(admin, hbaseTableId)) {
```
Comment (sagarkapare) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/hbase/HBaseStreamConsumerStateStoreFactory.java:65]: Since `tableExist` method is not part of the `HBaseDDLExecutor`, however we still want to reuse the `admin`, should `HBaseDDLExecutor` have `getAdmin` method?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/kv/HBaseKVTableDefinition.java [hunk 5]
```diff
@@ -121,8 +125,8 @@ public void create() throws IOException {
       HTableDescriptorBuilder tableDescriptor = tableUtil.buildHTableDescriptor(tableId);
       tableDescriptor.addFamily(columnDescriptor);
 
-      try (HBaseAdmin admin = new HBaseAdmin(hConf)) {
-        tableUtil.createTableIfNotExists(admin, tableId, tableDescriptor.build());
+      try (HBaseDDLExecutor executor = new HBaseDDLExecutorFactory(cConf, hConf).get()) {
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/kv/HBaseKVTableDefinition.java]: seems like you can create the factory in the constructor.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java [hunk 6]
```diff
@@ -552,8 +554,8 @@ private void createQueueTable(TableId tableId, HTableDescriptorBuilder htd, byte
       htd.setValue(HBaseQueueAdmin.PROPERTY_PREFIX_BYTES, Integer.toString(prefixBytes));
       LOG.info("Create queue table with prefix bytes {}", htd.getValue(HBaseQueueAdmin.PROPERTY_PREFIX_BYTES));
 
-      try (HBaseAdmin admin = new HBaseAdmin(hConf)) {
-        tableUtil.createTableIfNotExists(admin, tableId, htd.build(), splitKeys);
+      try (HBaseDDLExecutor executor = new HBaseDDLExecutorFactory(cConf, hConf).get()) {
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: can the factory be created once?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/hbase/HBaseStreamConsumerStateStoreFactory.java (hunk 4 above)
Comment (albertshau) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/hbase/HBaseStreamConsumerStateStoreFactory.java:65]: shouldn't have a getAdmin, as that breaks the abstraction. You can probably just always call createTableIfNotExists, unless it's expensive to build the table descriptor. Otherwise, it's probably ok to have the 2 calls, though I would only create the DDLExecutor if the table doesn't exist.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/hbase/HBaseStreamFileConsumerFactory.java [hunk 7]
```diff
@@ -88,8 +90,8 @@ protected StreamConsumer create(TableId tableId, StreamConfig streamConfig, Cons
     htd.addFamily(hcd);
     htd.setValue(QueueConstants.DISTRIBUTOR_BUCKETS, Integer.toString(splits));
 
-    try (HBaseAdmin admin = new HBaseAdmin(hConf)) {
-      tableUtil.createTableIfNotExists(admin, hBaseTableId, htd.build(), splitKeys,
+    try (HBaseDDLExecutor executor = new HBaseDDLExecutorFactory(cConf, hConf).get()) {
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/hbase/HBaseStreamFileConsumerFactory.java]: can the factory be created once in the constructor?

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/DefaultHBaseDDLExecutor.java [hunk 8]
```diff
@@ -45,14 +45,27 @@
...
+    try {
+      this.admin = new HBaseAdmin((Configuration) conf);
+    } catch (Exception e) {
+      throw new RuntimeException("Failed to create HBaseAdmin.", e);
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T> T getConf() {
```
Comment (albertshau) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/DefaultHBaseDDLExecutor.java:62]: why is this needed?

File: cdap-hbase-spi/src/main/java/co/cask/cdap/spi/hbase/HBaseDDLExecutor.java [hunk 9]
```diff
@@ -17,13 +17,29 @@
...
+   * @param <T>
+   */
+  <T> void initialize(T conf);
+
+  /**
+   * Returns the conf with which the {@link HBaseDDLExecutor} was initialized.
+   * @param <T>
+   * @return the conf
+   */
+  <T> T getConf();
```
Comment (albertshau) [inline cdap-hbase-spi/src/main/java/co/cask/cdap/spi/hbase/HBaseDDLExecutor.java:41]: why is this needed?

File: cdap-hbase-spi/src/main/java/co/cask/cdap/spi/hbase/HBaseDDLExecutor.java [hunk 10]
```diff
@@ -17,13 +17,29 @@
...
  */
-public interface HBaseDDLExecutor {
+public interface HBaseDDLExecutor extends Closeable {
+
+  /**
+   * Initialize the {@link HBaseDDLExecutor} with the configuration object provided
+   * @param conf the configurations with which to initialize
+   * @param <T>
+   */
+  <T> void initialize(T conf);
```
Comment (albertshau) [inline cdap-hbase-spi/src/main/java/co/cask/cdap/spi/hbase/HBaseDDLExecutor.java:34]: I think it's better to pass in some new Context object, as we can add methods to that Context without breaking this interface. That Context would have the getConfiguration() method. We can do this in a separate PR as well, when we add the ability to access configuration properties in the cconf.

**Review by albertshau — COMMENTED** [review]
a few minor comments

File: cdap-hbase-spi/src/main/java/co/cask/cdap/spi/hbase/HBaseDDLExecutor.java (hunk 9 above)
Comment (sagarkapare) [reply cdap-hbase-spi/src/main/java/co/cask/cdap/spi/hbase/HBaseDDLExecutor.java:41]: This is required because https://github.com/caskdata/cdap/blob/develop/cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java#L251 here we use conf to get the compression. But by looking at the comment on the method, seems like this is only used for test case though. Will see if can be removed in the next PR when we pass the `TableDescriptor`

File: cdap-hbase-spi/src/main/java/co/cask/cdap/spi/hbase/HBaseDDLExecutor.java (hunk 10 above)
Comment (sagarkapare) [reply cdap-hbase-spi/src/main/java/co/cask/cdap/spi/hbase/HBaseDDLExecutor.java:34]: Agree. Will do it as a separate PR.

**Comment by sagarkapare** [comment]
@albertshau Please take a look again.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DistributedStorageProviderNamespaceAdmin.java [hunk 11]
```diff
@@ -69,10 +73,11 @@ public void create(NamespaceMeta namespaceMeta) throws IOException, ExploreExcep
     // TODO: CDAP-1519: Create base directory for filesets under namespace home
     // create HBase namespace
     String hbaseNamespace = tableUtil.getHBaseNamespace(namespaceMeta);
-    try (HBaseAdmin admin = new HBaseAdmin(hConf)) {
+    try (HBaseDDLExecutor executor = hBaseDDLExecutorFactory.get();
+         HBaseAdmin admin = new HBaseAdmin(hConf)) {
       if (Strings.isNullOrEmpty(namespaceMeta.getConfig().getHbaseNamespace())) {
         try {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DistributedStorageProviderNamespaceAdmin.java:81]: should create the executor here instead of in the outside try

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DistributedStorageProviderNamespaceAdmin.java [hunk 12]
```diff
@@ -69,10 +73,11 @@ public void create(NamespaceMeta namespaceMeta) throws IOException, ExploreExcep
     // TODO: CDAP-1519: Create base directory for filesets under namespace home
     // create HBase namespace
     String hbaseNamespace = tableUtil.getHBaseNamespace(namespaceMeta);
-    try (HBaseAdmin admin = new HBaseAdmin(hConf)) {
+    try (HBaseDDLExecutor executor = hBaseDDLExecutorFactory.get();
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DistributedStorageProviderNamespaceAdmin.java]: remove this try and put it just where we use it to call HBaseAdmin.hasNamespace() to avoid creating unnecessarily

File: cdap-hbase-compat-0.96/src/main/java/co/cask/cdap/data2/util/hbase/HBase96TableUtil.java [hunk 13]
```diff
@@ -84,14 +85,10 @@ public boolean hasNamespace(HBaseAdmin admin, String namespace) throws IOExcepti
   }
 
   @Override
-  public void createNamespaceIfNotExists(HBaseAdmin admin, String namespace) throws IOException {
-    Preconditions.checkArgument(admin != null, "HBaseAdmin should not be null");
+  public void createNamespaceIfNotExists(HBaseDDLExecutor ddlExecutor, String namespace) throws IOException {
```
Comment (albertshau) [inline cdap-hbase-compat-0.96/src/main/java/co/cask/cdap/data2/util/hbase/HBase96TableUtil.java]: Can we remove this method and just call the ddlExecutor's method? It doesn't do anything at this point.

File: cdap-hbase-compat-0.96/src/main/java/co/cask/cdap/data2/util/hbase/HBase96TableUtil.java [hunk 14]
```diff
@@ -92,12 +91,10 @@ public void createNamespaceIfNotExists(HBaseDDLExecutor ddlExecutor, String name
   }
 
   @Override
-  public void deleteNamespaceIfExists(HBaseAdmin admin, String namespace) throws IOException {
-    Preconditions.checkArgument(admin != null, "HBaseAdmin should not be null");
+  public void deleteNamespaceIfExists(HBaseDDLExecutor ddlExecutor, String namespace) throws IOException {
```
Comment (albertshau) [inline cdap-hbase-compat-0.96/src/main/java/co/cask/cdap/data2/util/hbase/HBase96TableUtil.java]: similar comment, can we remove this in favor of just using the ddlExecutor to delete?

File: cdap-hbase-compat-1.2-cdh5.7.0/src/main/java/co/cask/cdap/data2/util/hbase/HBase12CDH570TableUtil.java [hunk 15]
```diff
@@ -102,17 +104,19 @@ public void deleteNamespaceIfExists(HBaseDDLExecutor ddlExecutor, String namespa
   }
 
   @Override
-  public void disableTable(HBaseAdmin admin, TableId tableId) throws IOException {
-    Preconditions.checkArgument(admin != null, "HBaseAdmin should not be null");
+  public void disableTable(HBaseDDLExecutor ddlExecutor, TableId tableId) throws IOException {
+    Preconditions.checkArgument(ddlExecutor != null, "HBaseDDLExecutor should not be null");
     Preconditions.checkArgument(tableId != null, "Table Id should not be null.");
-    admin.disableTable(nameConverter.toTableName(tablePrefix, tableId));
+    TableName tableName = nameConverter.toTableName(tablePrefix, tableId);
```
Comment (albertshau) [inline cdap-hbase-compat-1.2-cdh5.7.0/src/main/java/co/cask/cdap/data2/util/hbase/HBase12CDH570TableUtil.java]: just took a look at HTableNameConverter and I don't understand why its a concrete class and not a utility class with static methods. Can't disableTable and enableTable be the same for all versions of hbase? And if so, could we just remove those methods from HBaseTableUtil in favor of the ddl executor?

**Review by albertshau — COMMENTED** [review]
some questions

**Comment by sagarkapare** [comment]
@albertshau Addressed comments. Please take a look again. Thanks!

File: cdap-tms/src/main/java/co/cask/cdap/messaging/store/hbase/HBaseTableFactory.java [hunk 16]
```diff
@@ -169,9 +172,22 @@ public void disablePayloadTable(String tableName) throws IOException {
     }
   }
 
+  private void enableTable(HBaseDDLExecutor ddlExecutor, TableId tableId) throws IOException {
+    try {
+      TableName tableName = HTableNameConverter.toTableName(cConf.get(Constants.Dataset.TABLE_PREFIX), tableId);
+      ddlExecutor.enableTableIfDisabled(tableName.getNamespaceAsString(), tableName.getQualifierAsString());
+      LOG.debug("TMS Table {} has been disabled", tableName);
```
Comment (albertshau) [inline cdap-tms/src/main/java/co/cask/cdap/messaging/store/hbase/HBaseTableFactory.java]: disabled -> enabled

**Review by albertshau — APPROVED** [review]
one comment on a log message, other lgtm. fix and merge

**Comment by sagarkapare** [comment]
Resolved conflicts. Running build https://builds.cask.co/browse/CDAP-DUT5316-9

**Comment by sagarkapare** [comment]
Build is green. Squashing and merging.
